graphql/query_resolver.go

```diff
@@ -438,15 +438,27 @@ func (r *queryResolver) Project(ctx context.Context, projectIdentifier string) (
 
 // Projects is the resolver for the projects field.
 func (r *queryResolver) Projects(ctx context.Context) ([]*GroupedProjects, error) {
+	usr := mustHaveUser(ctx)
+	viewableProjects, err := usr.GetViewableProjectSettings(ctx)
+	if err != nil {
+		return nil, InternalServerError.Send(ctx, fmt.Sprintf("error getting viewable projects for '%s': '%s'", usr.DispName, err.Error()))
+	}
 	allProjects, err := model.FindAllMergedTrackedProjectRefs()
 	if err != nil {
 		return nil, ResourceNotFound.Send(ctx, err.Error())
 	}
 	// We have to iterate over the merged project refs to verify if they are enabled
 	enabledProjects := []model.ProjectRef{}
 	for _, p := range allProjects {
-		if p.Enabled {
+		if !p.Enabled {
+			continue
+		}
+		if !p.IsRestricted() {
 			enabledProjects = append(enabledProjects, p)
```

Comment:
Does this perhaps make sense to include in `GetViewableProjectSettings`? I kinda think the function name implies that permission checking is handled there (maybe that's how we ended up with this lapse in the first place 💀 )

Reply:
+1, I think instead of calling FindAllMergedTrackedProjectRefs we could just call FindMergedProjectRefsByIds and pass in the viewable projects list.

Reply:
I think its already handled by this call to [`FindAllowedResources`](https://github.com/evergreen-ci/evergreen/blob/89aa79229babc42d1a830fe42fd0f2f6b6cd97f5/model/user/user.go#L343) I believe it handles only fetching the projects a user is allowed to view.

Reply:
If `FindAllowedResources` already filters out projects that are restricted to the user then why would we need to add the additional `IsRestricted` check?

Reply:
I think the code here can probably be cleaned up a bit more or use some comments.
https://github.com/evergreen-ci/evergreen/blob/a7ac6092d5d4e03e321bc3479ebfe04c3aa02d6d/graphql/query_resolver.go#L442 fetches all of the projects a user can view.
https://github.com/evergreen-ci/evergreen/blob/a7ac6092d5d4e03e321bc3479ebfe04c3aa02d6d/graphql/query_resolver.go#L446 Then finds all of the projects that are not hidden but includes projects that are restricted since it [does not match on them](https://github.com/evergreen-ci/evergreen/blob/89aa79229babc42d1a830fe42fd0f2f6b6cd97f5/model/project_ref.go#L1348).
Then the for loop handles iterating through them and merging those arrays while also removing the restricted projects since they are fetched in the call to `FindAllMergedTrackedProjectRefs`.

> +1, I think instead of calling FindAllMergedTrackedProjectRefs we could just call FindMergedProjectRefsByIds and pass in the viewable projects list.

@ablack12 
This wouldn't be ideal since this query serves the projects dropdown on the project health page and we currently allow people to view all nonrestricted projects not just the ones users explicitly have permission to view.

Reply:
I think I will create a new db query that filters out restricted projects as well as hidden ones to avoid the step in the for loop.

Reply:
Ah, @khelif96 https://github.com/evergreen-ci/evergreen/blob/feee63f477d3f4d71d2183de391aae3a7fba9d91/globals.go#L1216 should be the permission that corresponds (users can see all projects that they have this permission for). I'd suggest making a "GetViewableProjects" or something.

Reply:
Just to reiterate, there's a world where we change permissions so that everyone by default CAN'T just see unrestricted projects. We should start relying entirely on the permissions model now, as opposed to making assumptions about project ref fields, so that we're prepared. https://jira.mongodb.org/browse/DEVPROD-5243

Reply:
Hey @ablack12  I just gave this a shot I think relying on permissions for this is the right way to go but up until DEVPROD-5243 is completed I don't think we can rely just on permissions. Relying just on `TasksView` Permission only allows the user to view the projects they are explicitly allowed to view so the projects dropdown would no longer behave the same way.